srml/babe/src/lib.rs

```diff
@@ -127,25 +161,57 @@ impl<T: Trait> Module<T> {
 		// the majority of their slot.
 		<timestamp::Module<T>>::minimum_period().saturating_mul(2.into())
 	}
-}
-
-impl<T: Trait> OnTimestampSet<T::Moment> for Module<T> {
-	fn on_timestamp_set(_moment: T::Moment) { }
-}
 
-impl<T: Trait> Module<T> {
 	fn change_authorities(new: Vec<AuthorityId>) {
 		Authorities::put(&new);
 
 		let log: DigestItem<T::Hash> = DigestItem::Consensus(BABE_ENGINE_ID, new.encode());
 		<system::Module<T>>::deposit_log(log.into());
 	}
+
+	fn deposit_vrf_output(vrf_output: &[u8; VRF_OUTPUT_LENGTH]) {
+		let l = Randomness::get();
+		let mut arr = [0u8; VRF_OUTPUT_LENGTH + 32];
+		arr[0..32].copy_from_slice(&l[..]);
+		arr[32..VRF_OUTPUT_LENGTH + 32].copy_from_slice(&vrf_output[..]);
+		Randomness::put(runtime_io::blake2_256(&arr));
+	}
+
+	pub fn process_inherent_digests() {
+		let mut is_first_babe_digest = true;
+		for i in Self::get_inherent_digests()
+			.logs
+			.iter()
+			.filter_map(|s| s.as_pre_runtime())
+			.filter_map(|(engine, mut data)| if engine == BABE_ENGINE_ID {
+				Decode::decode(&mut data)
+			} else { None }) {
+			assert!(is_first_babe_digest, "BABE only allows one BABE pre-digest; qed");
```

Comment:
favor `ensure!` in the runtime.

Reply:
This leads to compile-time errors.  These invariants are checked by the BABE consensus engine, though.